docs/relayer_hldd.org

```diff
@@ -0,0 +1,112 @@
+#+STARTUP: overview
+
+* DONE Introduction
+
+This document describes conceptual design of a relayer process, designed to listen to, sign and relay cross-chain transfer requests between Ethereum and AlephZero.
+
+System architecture can be divided into components responsible for listening to and handling requests coming from the AlephZero chain and into Ethereum and the ones responsible for handling cross chain transfers in the opposite direction.
+
+System is composed of various coupled components which comunicate with channels (akin to message passing, albeit much simpler).
+
+Becasue of the vast differences in how these two chains operate and finalize blocks there are some differences in how these two sides operate.
+
+* DONE Components handling AlephZero -> Ethereum transfer requests
+
+#+BEGIN_SRC plantuml :file azero_eth_components.png
+database "Cache" {
+  [RedisManager] as redis
+}
+
+package "AlephZero events" {
+  [Listener] as a0_listener
+  [Handler 1] as a0_handler_1
+  [...]
+  [Handler n] as a0_handler_n
+}
+
+package "Message channels" {
+  [Next block number] as a0_block_num
+  [Block seal] as a0_block_seal
+}
+
+cloud {
+  [Ethereum] as eth
+  [AlephZero] as a0
+}
+
+a0_listener <.up. a0  : events subscription
+a0_listener <.up. a0_block_num :  next block subscrption
+a0_listener .up.> a0_block_seal : publish last processed block
+redis .down.> a0_block_num : publish next block
+redis <.up. a0_block_seal : last processed block subscription
+
+a0_listener <.down.> a0_handler_1
+a0_listener <.down.> a0_handler_n
+
+a0_handler_1 .down.> eth : tx
+a0_handler_n .down.> eth : tx
+#+END_SRC
+[[file:azero_eth_components.png][azero_eth_components.png]]
+
+- *Listener*: maintains a subscription to the AlephZero chain. On boot it will receive the number of the next block to process and filter all the cross-chain transfer request events that occured between this cached number and the last finalized block. After that requests are handled concurrently in chunks of events from consecutive block ranges. Cache is updated as the concurrent event handlers spawned for each batch finish processing them, in the same order the batches arrived and were queued.
+- *Handlers*: independent event handlers that are spawned for each request. After all the events from the batch of blocks acknowledge they have finished cache gets updated with the next block number.
+Handlers sign and submit transactions to the MOST smart contract on the Ethereum chain, first by checking whether a given guardian signature is needed (maybe she already signed it, or maybe the request is already processed, or maybe some previous blocks were yanked). As long as it is needed in the last finalized block on ethereum, handler will sign it and wait for as long as the transaction appears in the last finalized block. Following pseudo-code is a conceptual representation of the exact logic followed by the handlers before acknowleding an cross-chain transfer request as handled:
```

Comment:
```suggestion
Handlers sign and submit transactions to the MOST smart contract on the Ethereum chain, first by checking whether a given guardian signature is needed (maybe they already signed it, or maybe the request is already processed, or maybe some previous blocks were yanked). As long as it is needed in the last finalized block on ethereum, the handler will sign it and wait until the transaction appears in the last finalized block. The following pseudo-code is a conceptual representation of the exact logic followed by the handlers before acknowledging a cross-chain transfer request as handled:
```

Reply:
there is a mistake in the description, see latest changes

Reply:
Have you pushed them yet? I don't seem to see anything new.

Reply:
:+1: